src/til/ut_til/RunLengthEncodingTests.cpp

```diff
@@ -0,0 +1,516 @@
+    }
+
+    static basic_container rle_decode(const rle_container& from)
+    {
+        basic_container to;
+        to.reserve(from.size());
+
+        for (const auto& run : from)
+        {
+            for (size_t i = 0; i < run.length; ++i)
+            {
+                to.push_back(run.value);
+            }
+        }
+
+        return to;
+    }
+
+    TEST_CLASS(RunLengthEncodingTests)
+
+    TEST_METHOD(ConstructDefault)
+    {
+        rle_vector rle{};
+        VERIFY_ARE_EQUAL(0u, rle.size());
+        VERIFY_IS_TRUE(rle.empty());
+
+        // We're testing replace() elsewhere, but this is special:
+        // This ensures that even if we're default constructed we can add data.
+        rle.replace(0, 0, { 1, 5 });
+        VERIFY_ARE_EQUAL(5u, rle.size());
+        VERIFY_IS_FALSE(rle.empty());
+    }
+
+    TEST_METHOD(ConstructWithInitializerList)
+    {
+        rle_vector rle{ { { 1, 3 }, { 2, 2 }, { 1, 3 } } };
+        VERIFY_ARE_EQUAL("1 1 1|2 2|1 1 1"sv, rle);
+    }
+
+    TEST_METHOD(ConstructWithLengthAndValue)
+    {
+        rle_vector rle(5, 1);
+        VERIFY_ARE_EQUAL("1 1 1 1 1"sv, rle);
+    }
+
+    TEST_METHOD(CopyAndMove)
+    {
+        constexpr auto expected_full = "1 1 1|2 2|1 1 1"sv;
+        constexpr auto expected_empty = ""sv;
+
+        rle_vector rle1{ { { 1, 3 }, { 2, 2 }, { 1, 3 } } };
+        rle_vector rle2;
+        VERIFY_ARE_EQUAL(expected_full, rle1);
+        VERIFY_ARE_EQUAL(expected_empty, rle2);
+
+        // swap
+        rle1.swap(rle2);
+        VERIFY_ARE_EQUAL(expected_empty, rle1);
+        VERIFY_ARE_EQUAL(expected_full, rle2);
+
+        // copy
+        rle1 = rle2;
+        VERIFY_ARE_EQUAL(expected_full, rle1);
+        VERIFY_ARE_EQUAL(expected_full, rle2);
+
+        // prepare rle1 for the upcoming move
+
+        // move
+        rle1 = { { { 1, 1 } } };
+        rle1 = std::move(rle2);
+        VERIFY_ARE_EQUAL(expected_full, rle1);
+    }
+
+    TEST_METHOD(At)
+    {
+        rle_vector rle{
+            {
+                { 1, 1 },
+                { 3, 2 },
+                { 2, 1 },
+                { 1, 3 },
+                { 5, 2 },
+            }
+        };
+
+        VERIFY_ARE_EQUAL(1u, rle.at(0));
+        VERIFY_ARE_EQUAL(3u, rle.at(1));
+        VERIFY_ARE_EQUAL(3u, rle.at(2));
+        VERIFY_ARE_EQUAL(2u, rle.at(3));
+        VERIFY_ARE_EQUAL(1u, rle.at(4));
+        VERIFY_ARE_EQUAL(1u, rle.at(5));
+        VERIFY_ARE_EQUAL(1u, rle.at(6));
+        VERIFY_ARE_EQUAL(5u, rle.at(7));
+        VERIFY_ARE_EQUAL(5u, rle.at(8));
+        VERIFY_THROWS(rle.at(9), std::out_of_range);
+    }
+
+    TEST_METHOD(Slice)
+    {
+        rle_vector rle{
+            {
+                { 1, 1 },
+                { 3, 2 },
+                { 2, 1 },
+                { 1, 3 },
+                { 5, 2 },
+            }
+        };
+
+        VERIFY_ARE_EQUAL("1|3 3|2|1 1 1|5 5"sv, rle);
+        // empty
+        VERIFY_ARE_EQUAL(""sv, rle.slice(0, 0)); // begin

```

Comment:
FYI, `VERIFY_ARE_EQUAL` takes a 3rd parameter which can be a string you want printed out in the log output so you can more easily tell which sub-test this is. (If you wanted to do that over the comments.)